cmd/medius/images/push.go

```diff
@@ -107,45 +108,33 @@ func NewPublishImagesCommand(options *common.Options) *cobra.Command {
 }
 
 func (b *buildAndPublish) Do(entry *common.Entry, timestamp time.Time) ([]string, error) {
-	metadata := entry.Artifact.Metadata()
-	artifactInfo, err := entry.Artifact.Inspect()
+	metadata := entry.Artifacts[0].Metadata()
+	artifactInfo, err := entry.Artifacts[0].Inspect()
 	if err != nil {
 		return nil, fmt.Errorf("error introspecting artifact %q: %v", metadata.Describe(), err)
 	}
-	b.Log.Infof("Remote artifact checksum: %q", artifactInfo.SHA256Sum)
 
-	imageSha, err := b.getImageSha(metadata.Describe())
+	images, artifacts, err := b.buildImages(entry)
 	if err != nil {
 		return nil, err
 	}
-	if imageSha == artifactInfo.SHA256Sum && !b.Options.PublishImagesOptions.ForceBuild {
-		b.Log.Info("Nothing to do.")
-		return nil, nil
-	}
-	if errors.Is(b.Ctx.Err(), context.Canceled) {
-		return nil, b.Ctx.Err()
-	}
-
-	b.Log.Infof("Rebuild needed, downloading %q ...", artifactInfo.DownloadURL)
-	file, err := b.getArtifact(artifactInfo)
-	if err != nil {
-		return nil, err
-	}
-	defer os.Remove(file)
+	defer cleanupArtifacts(artifacts)
 
-	b.Log.Info("Building containerdisk ...")
-	containerDisk, err := build.ContainerDisk(file, build.ContainerDiskConfig(artifactInfo.SHA256Sum, metadata.AdditionalLabels))
+	containerDiskIndex, err := build.ContainerDiskIndex(images)
 	if err != nil {
-		return nil, fmt.Errorf("error creating the containerdisk : %v", err)
-	}
-	if errors.Is(b.Ctx.Err(), context.Canceled) {
-		return nil, b.Ctx.Err()
+		return nil, fmt.Errorf("error creating the containerdisk index : %v", err)
 	}
 
 	names := prepareTags(timestamp, b.Options.PublishImagesOptions.TargetRegistry, entry, artifactInfo)
 	for _, name := range names {
-		if err := b.pushImage(containerDisk, name); err != nil {
-			return nil, err
+		if len(images) > 1 {
+			if err := b.pushImageIndex(containerDiskIndex, name); err != nil {
+				return nil, err
+			}
+		} else {
```

Comment:
Okay, it still needs the check for `len == 1` when there was nothing to update. Thinking about this... we might have an issue when one Artifact changed while the other didn't. If one artifact changed still the whole disk needs to be rebuilt.

/approve cancel

Reply:
I think checking the length of images won't be needed. From the beginning we could just push always image index, even if it contains single image manifest, it's not an issue.

Anyways, later it would contain multiple image manifests, then we won't need to push single image manifest and this check can be removed.

Reply:
Sorry but I don't agree. I would keep pushing the single image manifest if there's only a single image inside. But that is not the main issue here. In my comment above I was concerned that the `buildImages` func will silently skip artifacts if they did not change, which potentially could lead to containerdisks missing images.

Imagine you have two artifacts which belong to the same containerdisk, but only one of the artifacts changed. `buildImages` will only build the changed artifact and will replace the whole containerdisk which is now missing the artifact that did not change.

Reply:
> I would keep pushing the single image manifest if there's only a single image inside.

Back then we both agreed that this functionality should remain, but again, there is no issue with pushing image index with single image manifest inside manifests list, until we'll have multiple images inside that will work on amd64 and arm64 architectures.

From OCI Image Index Specification [1] - There is no statement "don't push an image index with single image manifest, this is not allowed and is bad pattern":
```
The image index is a higher-level manifest which points to specific image manifests,
ideal for one or more platforms.
```
[1] https://github.com/opencontainers/image-spec/blob/main/image-index.md#oci-image-index-specification

> Imagine you have two artifacts which belong to the same containerdisk, but only one of the artifacts changed. buildImages will only build the changed artifact and will replace the whole containerdisk which is now missing the artifact that did not change.

Okay. If I understand it correctly, I'll give an example of this issue:
```
Name | Last modified | Size
CentOS-7-x86_64-GenericCloud-1606.qcow2.xz | 2016-07-05 15:12 | 253M
CentOS-7-aarch64-GenericCloud-1606.qcow2.xz | 2016-07-06 20:22 | 271M
```
We can see that `aarch64` produced day after `x86_64` - Currently existing code in this PR will build only single image manifest, and because we have only one image built, it'll push single image manifest and not an image index of multiple image manifests.

To solve this issue, I'd suggest to have always image index in container registry and get rid of single image manifest approach (and code).

E.g. We can check if remote image index exists and it has manifests [1]:

a). If yes, then just mutate it by `AppendManifests` [2] to add a new image manifest (in our example above, it'll add `aarch64` to an existing `x86_64`) - Then push modified image index to remote registry.

b). If no, then build a new image index that will contain either one or more image manifests inside manifests list - Then push the new image index to remote registry.

[1] https://github.com/google/go-containerregistry/blob/v0.19.1/pkg/v1/remote/index.go#L47
[2] https://github.com/google/go-containerregistry/blob/v0.19.1/pkg/v1/mutate/mutate.go#L91C6-L96

What do you think?

Reply:
In the docker manifest specification it is mentioned:

> A manifest list is a list of image layers that is created by specifying one or more (ideally more than one) image names.

Note that we are using Docker manifests, not OCI manifests. I don't see an inherent issue in pushing single image manifests, but IMO it is cleaner to avoid it and also I don't think it costs us that much in terms of coding. All the pieces are already there.

About your proposed solution that suggests fetching remote manifests: For me it sounds to complex. We have the `common.Entry` and if we know we need to rebuild one of it's artifacts, we should be able to rebuild the whole disk resp. all other artifacts too. That way mostly changes to `buildImages` should be needed.

Reply:
Done, thanks. I've added a new `rebuildNeeded` to check if rebuild is needed - If yes, then we'll build all artifacts.